galley/pkg/config/processor/transforms/ingress/virtualService.go

```diff
@@ -0,0 +1,298 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain ingressAdapter copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ingress
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"k8s.io/api/extensions/v1beta1"
+	ingress "k8s.io/api/extensions/v1beta1"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/event"
+	"istio.io/istio/galley/pkg/config/processing"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+type virtualServiceXform struct {
+	*event.FnTransform
+
+	options processing.ProcessorOptions
+
+	mu sync.Mutex
+
+	ingresses map[resource.Name]*resource.Entry
+	vsByHost  map[string]*syntheticVirtualService
+}
+
+func newVirtualServiceXform(o processing.ProcessorOptions) event.Transformer {
+	xform := &virtualServiceXform{
+		options: o,
+	}
+	xform.FnTransform = event.NewFnTransform(
+		collection.Names{metadata.K8SExtensionsV1Beta1Ingresses},
+		collection.Names{metadata.IstioNetworkingV1Alpha3Virtualservices},
+		xform.start,
+		xform.stop,
+		xform.handle)
+
+	return xform
+}
+
+// Start implements processing.Transformer
+func (g *virtualServiceXform) start() {
+	g.vsByHost = make(map[string]*syntheticVirtualService)
+
+	g.ingresses = make(map[resource.Name]*resource.Entry)
+}
+
+// Stop implements processing.Transformer
+func (g *virtualServiceXform) stop() {
+	g.vsByHost = nil
+
+	g.ingresses = nil
+}
+
+// Handle implements event.Handler
+func (g *virtualServiceXform) handle(e event.Event, h event.Handler) {
+	if g.options.MeshConfig.IngressControllerMode == meshconfig.MeshConfig_OFF {
+		// short circuit and return
+		return
+	}
+
+	switch e.Kind {
+	case event.Added, event.Updated:
+		if !shouldProcessIngress(g.options.MeshConfig, e.Entry) {
+			scope.Debugf("virtualServiceXform: Skipping ingress event: %v", e)
+			return
+		}
+
+		g.processIngress(e.Entry, h)
+
+	case event.Deleted:
+		ing, exists := g.ingresses[e.Entry.Metadata.Name]
+		if exists {
+			g.removeIngress(ing, h)
+			delete(g.ingresses, e.Entry.Metadata.Name)
+		}
+
+	default:
+		scope.Errorf("virtualServiceXForm.handle: unknown event: %v", e)
```

Comment:
I really wish all your debug statements were geared towards customers rather than developers. In this case, as in many other places, instead of giving a function name and a raw error like "unknown event", it would be so much  better to have log messages which are meaningful to a customer with a clearer indication of how to solve the specific issue.

I suppose we'll eventually get to a point where we do a survey of our log output for usability...

The standard log package has options to enable displaying call stacks so that you get function names (--log_caller and --log_stacktrace_level), which can be controlled by scope.

Reply:
Change this one to error. Most of the method name additions are actually for debug messages and not the info. I tried to keep the info messages more relevant to the high-level operation but things slip through.